File: Solutions/116_Populating_Next_Right_Pointers_in_Each_Node/116_Populating_Next_Right_Pointers_in_Each_Node.go
//Second fastest
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Left *Node
 *     Right *Node
 *     Next *Node
 * }
 */

func connect(root *Node) *Node {
    if root == nil {
        return nil
    }
    queue := make([]*Node, 0)
    queue = append(queue, root)
    for len(queue) > 0 {
        l := len(queue) 
        for i := 0; i < l; i++ { //iterate l elements only
            node := queue[0]
            queue = queue[1:]
            
            if i == l - 1 {
                node.Next = nil //last node in this level
            } else if i < l - 1 {
                node.Next = queue[0] //not the last node in this level
            }
            
            if node.Left != nil {
                queue = append(queue, node.Left)
            }
            if node.Right != nil {
                queue = append(queue, node.Right)
            }
            
        }
    }
    return root
}

//Another
func connect(root *Node) *Node {
	
    itr := root
    for root != nil && root.Left != nil {
        start := root
        
        for start != nil {
            start.Left.Next = start.Right
            if start.Next != nil {
                start.Right.Next = start.Next.Left
            }
            start = start.Next
        }
        
        root = root.Left
    }
    
    return itr
}

//Fastest
func connectNext(node *Node) {
    if node == nil {
        return
    }
    if node.Next != nil {
        connectNext(node.Next)
    }
    
    if node.Left != nil {
        if node.Right != nil {
            node.Left.Next = node.Right
            node.Right.Next = getNextRight(node)
        } else {
            node.Left.Next = getNextRight(node)
        }
        connectNext(node.Left)
    } else if node.Right != nil {
        node.Right.Next = getNextRight(node)
        connectNext(node.Right)
    } else {
        connectNext(getNextRight(node))
    }
}

func getNextRight(node *Node) *Node {
    for nextRight := node.Next ; nextRight != nil; nextRight = nextRight.Next {
        if nextRight.Left != nil {
            return nextRight.Left
        } else if nextRight.Right != nil {
            return nextRight.Right
        }
    }
    return nil
}

func nextLevel(node *Node) *Node {
    if node.Left != nil {
        return node.Left
    } else if node.Right != nil {
        return node.Right
    } 
    return getNextRight(node)
}

func connectRight(root *Node) {
    for level := root; level != nil ; level = nextLevel(level) {
        for node := level; node != nil; node = node.Next {
            if node.Left != nil {
                if node.Right != nil {
                    node.Left.Next = node.Right
                } else {
                    node.Left.Next = getNextRight(node)
                }
            }
            if node.Right != nil {
                node.Right.Next = getNextRight(node)
            }
        }
    }
}

func connect(root *Node) *Node {
    connectRight(root)
    return root
}

File: Solutions/116_Populating_Next_Right_Pointers_in_Each_Node/116_Populating_Next_Right_Pointers_in_Each_Node.cs
//Most common
/*
// Definition for a Node.
public class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
}
*/

public class Solution {
    public Node Connect(Node root) {
        if(root == null)
            return null;
        var q = new Queue<Node>();
        q.Enqueue(root);
        while(q.Count != 0){
            int n = q.Count;
            for(int i=0; i<n; i++){
                var cur = q.Dequeue();
                if(i == n-1)
                    cur.next = null;
                else
                    cur.next = q.Peek();
                if(cur.left != null)
                    q.Enqueue(cur.left);
                if(cur.right != null)
                    q.Enqueue(cur.right);
            }
        }
        return root;
    }
}

//Another
public class Solution {
    public Node Connect(Node root) {
        if(root == null)
            return null;
        
        Queue<Node> queue = new Queue<Node>();
        queue.Enqueue(root);
        
        while(queue.Count > 0){
            int cnt = queue.Count;                
                for(int i = 0; i < cnt; i++){
                Node curr = queue.Dequeue();
                    if(i< cnt-1)
                        curr.next = queue.Peek();
                if(curr.left != null)
                    queue.Enqueue(curr.left);
                if(curr.right != null)
                    queue.Enqueue(curr.right);
            }
        }
        
        return root;
    }
}

//Third fastest
public class Solution {
    public Node Connect(Node root) {

        if(root == null)
            return root;
        
        var p1 = root;
        Node p2;
        
        
        while(p1.left != null)
        {
            p1.left.next = p1.right;
            p2 = p1;
            while(p2 != null)
            {
                // connect left and right
                p2.left.next = p2.right;
                
                if(p2.next != null)
                {
                    p2.right.next = p2.next.left;
                }
                p2 = p2.next;
            }
            p1 = p1.left;
        }
        
        return root;
    }
}

//Second fastest
public class Solution {
  public Node Connect(Node root) {
      
      if(root == null) return root;
      var q = new Queue<Node>();

      q.Enqueue(root);

      while(q.Count != 0)
      {

        Node prev = null;
        Node cur = null;
        int level = q.Count;
        for(int i = 1; i<= level; i++)
        {
          cur = q.Dequeue();
          if(cur.left != null) q.Enqueue(cur.left);
          if(cur.right != null) q.Enqueue(cur.right);


          if(prev != null)
          {
            prev.next = cur;
          }
          
          if(i == level) {
            cur.next = null;
            break;
          } 

          prev = cur;
        }
      }

      return root;
  }
}

//Fastest
public class Solution {
    public Node Connect(Node root) {
        var leftNode = root;
        while(leftNode!=null && leftNode.left!=null) {
            PopulateNode(leftNode);
            leftNode = leftNode.left;
        }
        return root;
    }
    private void PopulateNode(Node selectedNode) {
        var currentNode  = selectedNode;
        while(currentNode!=null) {
            currentNode.left.next = currentNode.right;
            if(currentNode.next!=null)
                currentNode.right.next = currentNode.next.left;
            currentNode = currentNode.next;
        }
    }
}